### PR TITLE
682 deprecated replacements

### PR DESCRIPTION
**Pull request by fenekku**
- closes galterlibrary/Prism#682

### CODE REVIEW

File: tests/contrib/lcsh/test_converter.py [hunk 1]
```diff
@@ -48,16 +49,65 @@ def test_deprecated():
     filepath = Path(__file__).parent / "data" / "fake_lcsh.skosrdf.jsonld"
     topics = list(read_jsonl(filepath))
 
-    dicts_of_replacements = [d for d in deprecated_to_replacements(topics)]
+    dicts_of_replacements = [d for d in raw_to_deprecated(topics)]
```
Comment (fenekku) [inline tests/contrib/lcsh/test_converter.py:52]: Entrypoint in the review: the changed tests and edge cases.

File: tests/contrib/lcsh/test_converter.py [hunk 2]
```diff
@@ -48,16 +49,65 @@ def test_deprecated():
...
+        # No clear new_id/new_subject
         {
             "id": "https://id.loc.gov/authorities/subjects/sh00000273",
             "time": "2021-07-20T08:25:20",
             "subject": "Child concentration camp inmates",
-            "new_id": "TODO:Expert:Fill me",
-            "new_subject": "TODO:Expert:Fill me",
+            "new_id": "",
+            "new_subject": "",
             "notes": "This authority record has been deleted because the heading is covered by the subject headings {Child internment camp inmates} (DLC)sh2021004026 and {Child Nazi concentration camp inmates} (DLC)sh2021004027"  # noqa
```
Comment (fenekku) [inline tests/contrib/lcsh/test_converter.py:62]: With Gretchen, we established to not associate any replacement id/subject in this case and let the curator manually select which should be more appropriate.

**Review by mattbaumann1 — APPROVED** [review]

File: galter_subjects_utils/contrib/lcsh/converter.py [hunk 3]
```diff
@@ -106,6 +109,23 @@ def is_deprecated(topic):
     return bool(find_deprecation_node(topic, node_main))
 
 
+def happened_since(topic, since=None):
+    """Filter for topic having happened after since.
+
+    :param topic: skos topic
+    :type since: dict
```
Comment (Meowcenary) [inline galter_subjects_utils/contrib/lcsh/converter.py:116]: I think this should be ":type topic"

File: galter_subjects_utils/contrib/lcsh/converter.py [hunk 4]
```diff
@@ -115,25 +135,47 @@ def find_explanation(nodes_of_change_note):
     return ""
 
 
+pattern = r"covered by .+ subject heading( {(?P<subject_new>.+)})? .+(?P<id_new>sh\s?\d+)$"  # noqa
+regex_for_replacement = re.compile(pattern)
```
Comment (Meowcenary) [inline galter_subjects_utils/contrib/lcsh/converter.py:139]: Why are these outside of a function? Is this being defined as a constant?

File: galter_subjects_utils/contrib/lcsh/converter.py [hunk 5]
```diff
@@ -115,25 +135,47 @@ def find_explanation(nodes_of_change_note):
...
 
 
-def deprecated_to_replacements(topics):
+def raw_to_deprecated(topics, since=None):
     """Iterator over deprecated topics and their relevant information.
 
     :param topics: LCSH Topics iterable
     :type topics: iterable[dict]
+    :param since: filter for this date (included) and later
+    :type topics: datetime.datetime
```
Comment (Meowcenary) [inline galter_subjects_utils/contrib/lcsh/converter.py:177]: ":type since: datetime.datetime"

File: galter_subjects_utils/contrib/lcsh/converter.py [hunk 6]
```diff
@@ -115,25 +135,47 @@ def find_explanation(nodes_of_change_note):
...
+    id_new = lcsh.generate_id(id_new) if id_new else id_new
+    subject_new = matches.group("subject_new") or ""
+    return id_new, subject_new
+
+
 def to_raw_replacement(topic):
     """Fill out replacement dict."""
     node_main = find_main_node(topic)
     node_deprecation = find_deprecation_node(topic, node_main)
+    notes = find_explanation(node_main.get("skos:changeNote"))
```
Comment (Meowcenary) [inline galter_subjects_utils/contrib/lcsh/converter.py:159]: Nice 👍

**Review by Meowcenary — APPROVED** [review]
This was a pleasure to read through! Really nice work here @fenekku 👏 I saw two small typos and had a quick question about why you placed your regex assignment where you did, but otherwise it looked good.

As a side note, I'm surprised to see you using a regular expression at all given your adage on them :P

**Review by sharpattack — APPROVED** [review]
Looks good to me, just a personal preference comment.

File: galter_subjects_utils/contrib/lcsh/converter.py (hunk 3 above)
Comment (fenekku) [reply galter_subjects_utils/contrib/lcsh/converter.py:116]: Yes, thanks!

File: galter_subjects_utils/contrib/lcsh/converter.py (hunk 4 above)
Comment (fenekku) [reply galter_subjects_utils/contrib/lcsh/converter.py:139]: These are outside the function to take advantage of regular expression compilation. Basically, `re.compile(pattern)` is a one-time hit in exchange for an efficient regular expression object that is more performant at executing the `search` operations than `re.search(<pattern>, <string>)`. *Technically* Python does regex compilation under the hood of `re.search` and caches the result such that this code pattern shouldn't matter if we are using few regexes. But given the size of the code base, I am assuming we have nothing to lose by performing this compilation.

Python also doesn't have enforced constants per se, but it would be the norm to capitalize these module level variables to indicate that they shouldn't be changed. I didn't here, because I didn't want them to standout/be seen as regular "constants" that could be reused elsewhere. Not very attached to that though, so I could change it if it's important for others.

File: galter_subjects_utils/contrib/lcsh/converter.py (hunk 5 above)
Comment (fenekku) [reply galter_subjects_utils/contrib/lcsh/converter.py:177]: Thanks!

File: galter_subjects_utils/contrib/lcsh/converter.py [hunk 7]
```diff
@@ -115,25 +135,47 @@ def find_explanation(nodes_of_change_note):
...
     return {
         "id": node_main.get("@id").replace("http://", "https://"),
         "time": node_deprecation.get("cs:createdDate", {}).get("@value"),
         "subject": node_main.get("skosxl:literalForm", {}).get("@value"),
-        "new_id": "TODO:Expert:Fill me",
-        "new_subject": "TODO:Expert:Fill me",
-        "notes": find_explanation(node_main.get("skos:changeNote")),
+        "new_id": id_new,
+        "new_subject": subject_new,
+        "notes": notes,
```
[the remark this answers is not in the document]
Comment (fenekku) [reply galter_subjects_utils/contrib/lcsh/converter.py:167]: Yeah... it was hard for me to find an order that made most sense. I also considered the following:

```
{
    "id": ...,
    "subject": ...,
    "new_id": ...,
    "new_subject": ...,
    "time": ...,
    "notes": ...
}
```

Since the pair `(id, subject)` (resp. `(new_id, new_subject)`) goes together as one entity and that reflects it better. 

In the end, I went with that order because it's the same as the one laid out in the CSV file, where the primary key is the `id` and the primary filtering is the `time`. I will keep this in mind and see if what works best for Gretchen.

File: galter_subjects_utils/contrib/lcsh/converter.py (hunk 4 above)
Comment (Meowcenary) [reply galter_subjects_utils/contrib/lcsh/converter.py:139]: Thanks for the explanation! I think it's fine as is. Just wanted to make sure I understood what was going on.
